### DISCUSSION TITLE
Is it possible to query structures on top of structures?

### DISCUSSION THREAD

**Discussion by Raj2032** [vincent-herlemont/native_db#195] [General]
THis seems like a very cool project and I like the fact that it is nosql and it integrates with Rust quite nicely and may be fit for my use case.

I just had a question.

Say I had this code:

```rust
struct Customer
{
    id: u16,
    name: Name,
    age: u8,
    is_active: bool,
}

struct Name
{
    first: String,
    last: String,
}

fn main()
{
    let customer = Customer{name: Name{first: "John".to_string(), last: "Doe".to_string()},
    age: 32, is_active: false};
}
```

I believe it is directly possible to insert from the `customer` variable after serializing it?

And more importantly how would I query for the first name in my exampel code as `name` is linked to another structure that contains `first` and `last` string data type fields?

**Comment by vincent-herlemont** [comment]
> I believe it is directly possible to insert from the customer variable after serializing it?

Native DB automatically handles calling Native Model to serialize your data. Read these links:
- [Quick Start](https://docs.rs/native_db/latest/native_db/#quick-start)
- For more details on [native_model](https://github.com/vincent-herlemont/native_model).

> And more importantly how would I query for the first name in my exampel code as name is linked to another structure that contains first and last string data type fields?

There are no constraints or foreign key principles like in SQL. It is a database that resembles a KV store with some additional functionalities. However, if you want to query your name as a secondary key, you can define your `Name` struct as a key. Explanation here: [ToKey](https://docs.rs/native_db/latest/native_db/db_type/trait.ToKey.html).

**Reply by Raj2032** [reply]
@vincent-herlemont Hey mate thanks for your response.

> Native DB automatically handles calling Native Model to serialize your data. Read these links:

That's great mate.

> There are no constraints or foreign key principles like in SQL. It is a database that resembles a KV store with some additional functionalities.

I see, but as `name` is wrapped in structure `Name`, I am not too sure how you would still query any rows with the name "John" without using secondary key, could I get a code example please?

**Reply by vincent-herlemont** [reply]
@Raj2032 Example follow up https://github.com/vincent-herlemont/native_db/discussions/207#discussioncomment-10228998
